x.c

```diff
@@ -1534,6 +1542,399 @@ xmakeglyphfontspecs(XftGlyphFontSpec *specs, const Glyph *glyphs, int len, int x
 	return numspecs;
 }
 
+static int
+issloperising (int x, int iPoint, int waveWidth)
+{
+	//    .     .     .     .
+	//   / \   / \   / \   / \
+	//  /   \ /   \ /   \ /   \
+	// .     .     .     .     .
+
+	// Find absolute `x` of point
+	x += iPoint * (waveWidth/2);
+
+	// Find index of absolute wave
+	int absSlope = x / ((float)waveWidth/2);
+
+	return (absSlope % 2);
+}
+
+static int
+getslope (int x, int iPoint, int waveWidth)
+{
+	// Sizes: Caps are half width of slopes
+	//    1_2       1_2       1_2      1_2
+	//   /   \     /   \     /   \    /   \
+	//  /     \   /     \   /     \  /     \
+	// 0       3_0       3_0      3_0       3_
+	// <2->    <1>         <---6---->
+
+	// Find type of first point
+	int firstType;
+	x -= (x / waveWidth) * waveWidth;
+	if (x < (waveWidth * (2.f/6.f)))
+		firstType = UNDERCURL_SLOPE_ASCENDING;
+	else if (x < (waveWidth * (3.f/6.f)))
+		firstType = UNDERCURL_SLOPE_TOP_CAP;
+	else if (x < (waveWidth * (5.f/6.f)))
+		firstType = UNDERCURL_SLOPE_DESCENDING;
+	else
+		firstType = UNDERCURL_SLOPE_BOTTOM_CAP;
+
+	// Find type of given point
+	int pointType = (iPoint % 4);
+	pointType += firstType;
+	pointType %= 4;
+
+	return pointType;
+}
+
+void
+drawundercurlcurly(GC ugc, int winx, int winy, int wlw, int width, int widthThreshold, int charlen)
+{
+	int ww = win.cw;//width;
+	int wh = dc.font.descent - wlw/2 - 1;//r.height/7;
+	int wx = winx;
+	int wy = winy + win.cyo + win.ch - dc.font.descent;
+	// Draw waves
+	int narcs = charlen * 2 + 1;
+	XArc *arcs = xmalloc(sizeof(XArc) * narcs);
+
+	int i = 0;
+	for (i = 0; i < charlen-1; i++) {
+		arcs[i*2] = (XArc) {
+			.x = wx + win.cw * i + ww / 4,
+			.y = wy,
+			.width = win.cw / 2,
+			.height = wh,
+			.angle1 = 0,
+			.angle2 = 180 * 64
+		};
+		arcs[i*2+1] = (XArc) {
+			.x = wx + win.cw * i + ww * 0.75,
+			.y = wy,
+			.width = win.cw/2,
+			.height = wh,
+			.angle1 = 180 * 64,
+			.angle2 = 180 * 64
+		};
+	}
+	// Last wave
+	arcs[i*2] = (XArc) {wx + ww * i + ww / 4, wy, ww / 2, wh,
+		0, 180 * 64 };
+	// Last wave tail
+	arcs[i*2+1] = (XArc) {wx + ww * i + ww * 0.75, wy, ceil(ww / 2.),
```

Comment:
you sure you've tried to replace both ceils to rounds here...

Reply:
nope, actually for first and last `ceil` makes sense, to avoid 'overflowing' from the width